src/NewButton/button.tsx

```diff
@@ -0,0 +1,220 @@
+      &:focus:not([disabled]) {
+        border-color: ${get('colors.btn.danger.focusBorder')({theme})};
+        box-shadow: ${get('shadows.btn.danger.focusShadow')({theme})};
+      }
+
+      &:active:not([disabled]) {
+        color: ${get('colors.btn.danger.selectedText')({theme})};
+        background-color: ${get('colors.btn.danger.selectedBg')({theme})};
+        box-shadow: ${get('shadows.btn.danger.selectedShadow')({theme})};
+        border-color: ${get('colors.btn.danger.selectedBorder')({theme})};
+      }
+
+      &:disabled {
+        color: ${get('colors.btn.danger.disabledText')({theme})};
+        background-color: ${get('colors.btn.danger.disabledBg')({theme})};
+        border-color: ${get('colors.btn.danger.disabledBorder')({theme})};
+      }
+    `,
+    invisible: `
+      color: ${get('colors.accent.fg')({theme})};
+      background-color: transparent;
+      border: 0;
+      border-radius: ${get('radii.2')({theme})};
+      box-shadow: none;
+    
+      &:disabled {
+        color: ${get('colors.primer.fg.disabled')({theme})};
+      }
+      &:focus:not([disabled]) {
+        box-shadow: ${get('shadows.btn.focusShadow')({theme})};
+      }
+      &:hover:not([disabled]) {
+        background-color: ${get('colors.btn.hoverBg')({theme})};
+      }
+      &:active:not([disabled]) {
+        background-color: ${get('colors.btn.selectedBg')({theme})};
+      }
+    `
+  }
+  return style[variant]
+}
+
+const getSizes = (size = 'medium', variant: VariantType = 'default', iconOnly: boolean) => {
+  let paddingTop, paddingLeft
+  switch (size) {
+    case 'small':
+      paddingTop = 3
+      paddingLeft = 12
+      break
+    case 'large':
+      paddingTop = 9
+      paddingLeft = 20
+      break
+    case 'medium':
+    default:
+      paddingTop = 5
+      paddingLeft = 16
+  }
+  if (iconOnly) {
+    paddingLeft = paddingTop + 2
+  }
+  if (variant === 'invisible') {
+    paddingTop = paddingTop + 1
+  }
+  return `
+    padding:${paddingTop}px ${paddingLeft}px;
+  `
+}
+const ButtonBase = styled.button<ButtonBaseProps & {iconOnly: boolean}>`
+  ${buttonBaseStyles}
+  ${props => getVariantStyles(props.theme, props.variant)}
+  ${props => getSizes(props.size, props.variant, props.iconOnly)}
+  ${sizes}
+  ${sx}
+  ${fontSize}
+`
+
+const Button = forwardRef<HTMLAnchorElement | HTMLButtonElement, ComponentProps<typeof ButtonBase>>(
+  ({children, ...props}, forwardedRef) => {
+    const {icon: Icon, caret, size} = props
+    let iconOnly = false
+    if (Icon && !children) {
+      iconOnly = true
+    }
+    const iconWrapStyles = {
+      display: 'inline-block',
+      ...(!iconOnly ? {mr: 2} : {})
+    }
+    return (
+      <ButtonBase ref={forwardedRef} {...props} iconOnly={iconOnly}>
+        {Icon && (
+          <Box sx={iconWrapStyles} aria-hidden={!iconOnly}>
+            <Icon size={size} />
+          </Box>
+        )}
+        {children}
+        {caret && (
```

Comment:
I'm wondering if this could be more generic, though perhaps we don't have a use case for a `trailingIcon` in button other than a dropdown menu? My instinct is that button should support `leading` and `trailing` icons (generic slots) and the Dropdown component should utilize `trailingIcon` for `<TriangleDownIcon>`.

Reply:
in the button audit anatomy here, there was no trailing icon specifically. https://github.com/github/primer/issues/253#issuecomment-949215306

Also there are special margin recommendations for the caret icon which wouldn't necessarily apply for a `trailingIcon`. Which is why I opted for this API.

Reply:
Thanks for linking that audit! The anatomy graphic Ash provided makes a lot of sense. I still think we could make this more generic, and maybe borrow some concepts from ActionList. Here's how I see Button based on that audit:

![image](https://user-images.githubusercontent.com/18661030/139874402-dffa20a9-20e1-4222-a20c-e927ded73925.png)

I'm also new to the team so I don't know if we have guidelines around this, but I've had success in the past by separating concerns. IMO `Button` shouldn't handle `Dropdown` specific controls- it should provide the tools that `Dropdown` needs to be a Dropdown (a slot for a caret, anything else necessary). 

Now that I've seen the audit, `counter` could be a `trailingVisual` or an action if its interactive in some way (I don't think it is?). I'm not sure how that works with with `Button.Counter` format, I'm not super familiar with dot notation. 

For the spacing comment: I think the audit is pointing out spacing inconsistencies (using whitespace instead of actual layout properties) and that could be handled in `trailingAction` even if its not "caret" specific.

@ashygee any thoughts here?

Reply:
@langermank I think that we could make the icon slot align to the `leadingVisual` naming convention for consistency. I can update that in my audit as it does make sense in the off chance we decide to allow for other leading visual accessories outside of octicons. The same for `label` too. For `trailingVisual` I could see us being more generic here but I would want to be explicit about what we would want to allow, at this point I think it would just be a counter or an octicon. 

I am hesitant about the `trailingAction` since that is primarily for dropdown carets only, I did not come across any additional actions in my original audit and can't imagine at this point what else we may want to pass.

@pksjce in my original audit I was just pointing out the different elements in the anatomy and not clearly recommending that those be the slot names so I apologize if that misled you. I can update the imagery with @langermank 's recommendation for clarity.

Reply:
Thanks @ashygee! There's always a fine line between building what you need and what you might need in the future 😆 I think for the caret dropdown, we can be both "future-proof" (ish) and refrain from building too much logic without a use case by handling that second visual/action slot as such (keep the naming generic) while _also_ shipping Button with the caret functionality IN that slot. So essentially a private API, but it will allow for us to expand upon it later should a new use case come up. For example...

```
<Button>
{trailingAction && (<Caret />)}
</Button>

// which could eventually become

<Button>
{trailingAction && dropdownMenu ? (<Caret />) : trailingActionSlot}
</Button>
```

So offer `trailingAction` with a default for now, but allowing for flexibility later. Another idea would be to leave it completely open now, but treat it as a private API and only use it in other Primer components- basically just don't document that its an available slot.

```
// DropdownMenu

const DropdownMenu = (
<Button trailingAction="caret" />
<Menu>
)
```

Reply:
I like what @siddharthkp did with his composable ActionList API. I'd recommend something like...

1. Specify `Button.Count` as it's own slot
```jsx
<Button>
    <Button.LeadingVisual><EyeIcon /><Button.LeadingVisual>
    Watch
    <Button.Count>42</Button.Count>
    <Button.TrailingVisual><CaretIcon /></Button.TrailingVisual>
</Button>
```

2. style `Button.TrailingVisual` with `display: flex` and allow multiple children:
```jsx
<Button>
    <Button.LeadingVisual><EyeIcon /><Button.LeadingVisual>
    Watch
    <Button.TrailingVisual>
        <CaretIcon />
        <CountLabel>42</CountLabel>
    </Button.TrailingVisual>
</Button>
```

Reply:
> style Button.TrailingVisual with display: flex and allow multiple children

I love this idea! Its just flexible enough, and we can still control the layout within the slot 😍